Add support for multi cluster setup in helm (#2169)
flyteorg/flyte: rsts/deployment/multicluster.rst

      apiVersion: v1
      kind: Secret
      metadata:
        name: cluster_credentials
        namespace: flyte
      type: Opaque
      data:
        cluster_1_token: {{ cluster 1 token here }}
        cluster_1_cacert: {{ cluster 1 cacert here }}
        cluster_2_token: {{ cluster 2 token here }}
        cluster_2_cacert: {{ cluster 2 cacert here }}
        cluster_3_token: {{ cluster 3 token here }}
        cluster_3_cacert: {{ cluster 3 cacert here }}

* Create cluster credentials secret in the control plane cluster

.. code-block::

    kubectl apply -f secrets.yaml

* Create a file named ``values-override.yaml`` and add the following config to it: ::

      flyteadmin:
        additionalVolumes:
        - name: cluster_credentials
          secret:
            secretName: cluster_credentials
        additionalVolumeMounts:
        - name: cluster_credentials
          mountPath: /var/run/credentials
      configmap:
        labelClusterMap:
          team1:
            - id: cluster_1
              weight: 1
          team2:
            - id: cluster_2
              weight: 0.5
            - id: cluster_3
              weight: 0.5

>         clusterConfigs:

Author: Fix the values

            type: "file_path"
            tokenPath: "/var/run/credentials/cluster_1_token"
            certPath: "/var/run/credentials/cluster_1_cacert"
        - name: "cluster_2"
          endpoint: {{ your-cluster-2-kubeapi-endpoint.com }}
          auth:
            enabled: true
            type: "file_path"
            tokenPath: "/var/run/credentials/cluster_2_token"
            certPath: "/var/run/credentials/cluster_2_cacert"
        - name: "cluster_3"
          endpoint: {{ your-cluster-3-kubeapi-endpoint.com }}
          enabled: true
          auth:
            type: "file_path"
            tokenPath: "/var/run/credentials/cluster_3_token"
            certPath: "/var/run/credentials/cluster_3_cacert"

* Install Flyte control plane Helm chart

.. tabbed:: AWS

    .. code-block::

        helm upgrade flyte -n flyte flyteorg/flyte-core values.yaml -f values-aws.yaml -f values-controlplane.yaml -f values-override.yaml --create-namespace flyte --install

.. tabbed:: GCP

    .. code-block::

        helm upgrade flyte -n flyte flyteorg/flyte-core values.yaml -f values-gcp.yaml -f values-controlplane.yaml -f values-override.yaml --create-namespace flyte --install

> Configure Execution Cluster Labels

Reviewer: Added the content.

-------------------------

.. tip::
  As described in the `Architecture Overview <https://docs.flyte.org/en/latest/concepts/architecture.html>`_, the Flyte ``Control Plane`` sends workflows off to the ``Data Plane`` for execution. The data plane fulfills these workflows by launching pods in Kubernetes.

Often, the total compute needs could exceed the limits of a single Kubernetes cluster.
To address this, you can deploy the data plane to several isolated Kubernetes clusters.

> The control-plane (FlyteAdmin) can be configured to load-balance workflows across these isolated data planes, which protects you from failure in a single Kubernetes cluster, increasing scalability.

Reviewer 2: Is this control plane or control-plane?

Reviewer 2: ```suggestion
The control plane (FlyteAdmin) can be configured to load-balance workflows across these isolated data planes, protecting you from failure in a single Kubernetes cluster increasing scalability.
```

Reviewer: I think it has to be "control plane". Let's use this for now.

Also, let's make the following sentence clear:

... protecting you from failure in a single Kubernetes cluster, thus increasing scalability.

The control-plane (FlyteAdmin) can be configured to load-balance workflows across these isolated data-planes, which protects you from failure in a single Kubernetes cluster, increasing scalability.
Often, the total compute needs could exceed the limits of a single Kubernetes cluster.
To address this, you can deploy the data plane to several isolated Kubernetes clusters.
The control-plane (FlyteAdmin) can be configured to load-balance workflows across these isolated data planes, which protects you from failure in a single Kubernetes cluster, increasing scalability.

To achieve this, first, you have to create additional Kubernetes clusters. 

> For now, let’s assume you have three Kubernetes clusters and that you can access them all with ``kubectl``. Let’s call these clusters ``cluster1``, ``cluster2``, and ``cluster3``.

Reviewer 2: Single quotes are curly here and in the next line they are straight.

> Lets call these clusters ``cluster1``, ``cluster2``, and ``cluster3``.

Reviewer 2: ```Suggestion
Let's call these clusters ``cluster1``, ``cluster2``, and ``cluster3``.
```

Data Plane Deployment
*********************

* Add Flyteorg Helm repo

.. code-block::

    helm repo add flytyteorg https://flyteorg.github.io/flyte
    helm repo update
    # Get flyte-core helm chart
    helm fetch --untar --untardir . flyteorg/flyte-core
    cd flyte-core

* Install Flyte data plane Helm chart

.. tabbed:: AWS

    .. code-block::

        helm upgrade flyte -n flyte flyteorg/flyte-core values.yaml -f values-gcp.yaml -f values-dataplane.yaml --create-namespace flyte --install

.. tabbed:: GCP

    .. code-block::

        helm upgrade flyte -n flyte flyteorg/flyte-core values.yaml -f values-aws.yaml -f values-dataplane.yaml --create-namespace flyte --install

User and Control Plane Deployment
*********************************

Some deployments of Flyte may choose to run control plane separate from the data plane.
FlyteAdmin is designed to create Kubernetes resources in one or more Flyte data plane clusters.
For the admin to access remote clusters, it needs credentials to each cluster.

In Kubernetes, scoped service credentials are created by configuring a “Role” resource in a Kubernetes cluster.

> When you attach the role to a “ServiceAccount”, Kubernetes generates a bearer token that permits access.

Reviewer 2: ```suggestion
When you attach the role to a "ServiceAccount", Kubernetes generates a bearer token that permits access.
```

Hence, create a FlyteAdmin `ServiceAccount <https://github.com/flyteorg/flyte/blob/master/charts/flyte-core/templates/admin/rbac.yaml#L4>`_ in each data plane cluster to generate these tokens.

> When you first create the FlyteAdmin ``ServiceAccount`` in a new cluster, a bearer token is generated and will continue to allow access unless the “ServiceAccount“ is deleted.

Reviewer 2: ```suggestion
When you first create the FlyteAdmin ``ServiceAccount`` in a new cluster, a bearer token is generated and will continue to allow access unless the "ServiceAccount" is deleted.
```

Hence, you should never delete a ``ServiceAccount`` ⚠️.

To feed the credentials to FlyteAdmin, you must retrieve them from your new data plane cluster, and upload them to admin somehow (for example, within Lyft, :ref:`Confidant <https://github.com/lyft/confidant>`__ is used).

> The credentials have two parts (“ca cert“ and “bearer token"). Find the generated secret via, ::

Reviewer 2: ```suggestion
The credentials have two parts ("ca cert" and "bearer token"). Find the generated secret via,::
```

User and Control Plane Deployment
*********************************

> Some deployments of Flyte may choose to run control plane separate from the data plane.

Reviewer 2: ```suggestion
Some Flyte deployments may choose to run the control plane separate from the data plane.
```

FlyteAdmin is designed to create Kubernetes resources in one or more Flyte data plane clusters.
For the admin to access remote clusters, it needs credentials to each cluster.

> In Kubernetes, scoped service credentials are created by configuring a “Role” resource in a Kubernetes cluster.

Reviewer 2: Please check the curly vs. straight quotes consistency.

When you attach the role to a “ServiceAccount”, Kubernetes generates a bearer token that permits access.
Hence, create a FlyteAdmin `ServiceAccount <https://github.com/flyteorg/flyte/blob/master/charts/flyte-core/templates/admin/rbac.yaml#L4>`_ in each data plane cluster to generate these tokens.

When you first create the FlyteAdmin ``ServiceAccount`` in a new cluster, a bearer token is generated and will continue to allow access unless the “ServiceAccount“ is deleted.
Hence, you should never delete a ``ServiceAccount`` ⚠️.

> To feed the credentials to FlyteAdmin, you must retrieve them from your new data plane cluster, and upload them to admin somehow (for example, within Lyft, :ref:`Confidant <https://github.com/lyft/confidant>`__ is used).

Reviewer 2: This sounds informal. Could we remove somehow?

> To feed the credentials to FlyteAdmin, you must retrieve them from your new data plane cluster, and upload them to admin somehow (for example, within Lyft, :ref:`Confidant <https://github.com/lyft/confidant>`__ is used).

Reviewer 2: ```suggestion
To feed the credentials to FlyteAdmin, you must retrieve them from your new data plane cluster and upload them to admin (for example, within Lyft, ref:`Confidant <https://github.com/lyft/confidant>`__ is used).
```
